### PR TITLE
fix chunking of attention mask

### PR DESCRIPTION
**Pull request by jglaser**
fixes a typo

### CODE REVIEW

**Review by hypnopump — APPROVED** [review]
Good catch!

**Comment by hypnopump** [comment]
btw, ive got some ideas (and prototypes) for greatly improving the speed of the pytorch implementation... the only downside is they would need a refactor of the code and depend on a nightly build of pytorch. Would you like to explore them?

**Comment by jglaser** [comment]
I'd be happy to look at any prototype code if it is available in a branch.... This attention layer seems to work as advertised, it reduces mem usage significantly allowing for a larger batch size during training.

**Comment by hypnopump** [comment]
hmm thats good! I have some tests of different components in a colab notebook but im not with the computer now, will link tmrw or so. 
One of the things im concerned about is speed. Benchmarks show actually a slower exe cution time than the standard implementation. Have you experienced significnat slow downs? 
Also, i also noticed the output values not to be exact (probably coming from numerical stability and the iterative computation of the max + the correction). Do you consider the models trained with this variant to perform worse/slightly worse? Also, could your try to see if just changing the attention kernel **but not the weights** causes a drop in inference performance? (A drop-in for inference would be quite useful)

**Comment by jglaser** [comment]
> hmm thats good! I have some tests of different components in a colab notebook but im not with the computer now, will link tmrw or so. One of the things im concerned about is speed. Benchmarks show actually a slower exe cution time than the standard implementation. Have you experienced significnat slow downs?

Not yet, it seems to be roughly on par with the standard attention, but I haven't benchmarked exhaustively. That said, I am always open for performance improvements.

> Also, i also noticed the output values not to be exact (probably coming from numerical stability and the iterative computation of the max + the correction). Do you consider the models trained with this variant to perform worse/slightly worse?

I found that it is exact up to numerical precision. I used the below script to confirm that:

```
import torch
import torch.nn as nn

from linear_mem_attention_pytorch.fast_attn import Attention

def transpose_for_scores(x, num_attention_heads, attention_head_size):
    new_x_shape = x.size()[:-1] + (num_attention_heads, attention_head_size)
    x = x.view(*new_x_shape)
    return x.permute(0, 2, 1, 3)

def bert_self_attention(query, key, value_layer, attention_mask=None, num_attention_heads=1):
        dim_head = query.size()[-1] // num_attention_heads
        all_head_size = dim_head*num_attention_heads

        query_layer = transpose_for_scores(query, num_attention_heads, dim_head)
        key_layer = transpose_for_scores(key, num_attention_heads, dim_head)
        value_layer = transpose_for_scores(value, num_attention_heads, dim_head)

        # Take the dot product between "query" and "key" to get the raw attention scores.
        attention_scores = torch.matmul(query_layer, key_layer.transpose(-1, -2))

        attention_scores = attention_scores / math.sqrt(dim_head)

        if attention_mask is not None:
            # Apply the attention mask is (precomputed for all layers in BertModel forward() function)
            attention_scores = attention_scores + attention_mask

        # Normalize the attention scores to probabilities.
        attention_probs = nn.functional.softmax(attention_scores, dim=-1)

        # This is actually dropping out entire tokens to attend to, which might
        # seem a bit unusual, but is taken from the original Transformer paper.
        #attention_probs = self.dropout(attention_probs)

        context_layer = torch.matmul(attention_probs, value_layer)

        context_layer = context_layer.permute(0, 2, 1, 3).contiguous()
        new_context_layer_shape = context_layer.size()[:-2] + (all_head_size,)
        context_layer = context_layer.view(*new_context_layer_shape)

        return context_layer, attention_probs

if __name__ == "__main__":
    query = torch.rand(1,200,2)
    key = value = query

    n_heads = 1
    attn, probs = bert_self_attention(query, key, value, num_attention_heads=n_heads)
    print('bert_self_attention scores: ', probs)
    print('bert_self_attention out: ', attn)

    dim_head = query.size()[-1]//n_heads
    linear_mem_attn = Attention(
        dim=query.size()[-1],
        heads=n_heads,
        dim_head=dim_head
    )

    linear_mem_attn.to_q = torch.nn.Identity()
    linear_mem_attn.to_kv = torch.nn.Identity()

    kv = torch.stack([key, value], dim=2)
    kv = torch.flatten(kv, start_dim=2)

    l_attn = linear_mem_attn(x=query, context=kv)
    print("Linear mem attention epsilon = {}".format(torch.linalg.matrix_norm(l_attn-attn).squeeze().numpy()))
```

The epsilon reported should be comparable to machine eps (1e-6 ish).

> Also, could your try to see if just changing the attention kernel **but not the weights** causes a drop in inference performance? (A drop-in for inference would be quite useful)

I haven't studied that systematically, but I did implement this kernel as a drop-in with pre-trained weights in this development code:
https://github.com/jglaser/affinity_pred/blob/10b005f6b2dbada1097d7dd22586ceb46c44cc77/affinity_pred/ensemble_embedding.py#L113-L183

So far, I am finding that it performs just as expected.

**Comment by jglaser** [comment]
I checked the execution time on an AMD GPU with 32GB RAM (gfx90a). I set the per-device batch size 1 for lowest latency. You're right, the default attention is faster in absolute terms. In my experiment with the above mentioned code used for fine tuning I find that the relative speedup is about ~30% (for the entire model). That said, the memory advantage still holds, as it allows me to train on older GPUs with less (16GB) memory, when the model with default attention does not fit into memory.

**Comment by hypnopump** [comment]
Oh crap i completely forgot hahah ok so my ideas for speeding this up involved: 
* V0: `torch.jit.script` but the current structure of nested functions is not very friendly for that, plus it only compiles python code to torch code. Problem is it didnt really offer that much of a speedup (like 10-20%)

* V1: Operation Authoring (an experimental torch feature that could accelerate it like 5x or so: https://www.youtube.com/watch?v=xsDhgn4mHLc ) I've already done some tests which indicate that forward pass is much more faster (4x-6x), but currently it only works on forward mode since our functions are a bit special and not all outputs require gradients: https://github.com/pytorch/functorch/issues/373
